jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/params/DefaultXGBoostParamsWriter.scala

```diff
@@ -0,0 +1,88 @@
+/*
+ Copyright (c) 2014 by Contributors
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+package ml.dmlc.xgboost4j.scala.spark.params
+
+import org.apache.hadoop.fs.Path
+import org.apache.spark.SparkContext
+import org.apache.spark.ml.param.{ParamPair, Params}
+import org.json4s.JsonDSL._
+import org.json4s.jackson.JsonMethods._
+import org.json4s.{JObject, _}
+
+// This originates from spark DefaultXGBoostParamsWriter & DefaultParamsReader copy paste
+private[spark] object DefaultXGBoostParamsWriter {
+
+  /**
+   * Saves metadata + Params to: path + "/metadata"
+   *  - class
+   *  - timestamp
+   *  - sparkVersion
+   *  - uid
+   *  - paramMap
+   *  - (optionally, extra metadata)
+   *
+   * @param extraMetadata Extra metadata to be saved at same level as uid, paramMap, etc.
+   * @param paramMap      If given, this is saved in the "paramMap" field.
+   *                      Otherwise, all [[org.apache.spark.ml.param.Param]]s are encoded using
+   *                      [[org.apache.spark.ml.param.Param.jsonEncode()]].
+   */
+  def saveMetadata(
+    instance: Params,
+    path: String,
+    sc: SparkContext,
+    extraMetadata: Option[JObject] = None,
+    paramMap: Option[JValue] = None
+  ): Unit = {
+    val metadataPath = new Path(path, "metadata").toString
+    val metadataJson = getMetadataToSave(instance, sc, extraMetadata, paramMap)
+    sc.parallelize(Seq(metadataJson), 1).saveAsTextFile(metadataPath)
```

Comment:
As I took it from https://github.com/apache/spark/blob/04ee8cf633e17b6bf95225a8dd77bf2e06980eb3/mllib/src/main/scala/org/apache/spark/ml/util/ReadWrite.scala#L275 I used their implementation. I think the idea is to have the model in a single file which is not spread over partitions (which is a nice assumption for small models i.e. tree based models).